plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,483 @@
+		}
+		if tags["temp_path"], ok = tempSpace["path"].(string); !ok {
+			return badFormatErr(te, tempSpace["path"], "string", tempSpaceMonitorRoute+".path")
+		}
+		if fields["temp_available"], ok = tempSpace["size"].(float64); !ok {
+			return badFormatErr(te, tempSpace["size"], "float64", tempSpaceMonitorRoute+".size")
+		}
+	}
+
+	if swapSpaceMonitor := info.MonitorData.Hudson_NodeMonitors_SwapSpaceMonitor; swapSpaceMonitor != nil {
+		swapSpaceMonitorRouter := "hudson.node_monitors.SwapSpaceMonitor"
+		var swapSpace map[string]interface{}
+		if swapSpace, ok = swapSpaceMonitor.(map[string]interface{}); !ok {
+			return badFormatErr(te, swapSpaceMonitor, "map[string]interface{}", swapSpaceMonitorRouter)
+		}
+		if fields["swap_available"], ok = swapSpace["availableSwapSpace"].(float64); !ok {
+			return badFormatErr(te, swapSpace["availableSwapSpace"], "float64", swapSpaceMonitorRouter+".availableSwapSpace")
+		}
+		if fields["swap_total"], ok = swapSpace["totalSwapSpace"].(float64); !ok {
+			return badFormatErr(te, swapSpace["totalSwapSpace"], "float64", swapSpaceMonitorRouter+".totalSwapSpace")
+		}
+		if fields["memory_available"], ok = swapSpace["availablePhysicalMemory"].(float64); !ok {
+			return badFormatErr(te, swapSpace["availablePhysicalMemory"], "float64", swapSpaceMonitorRouter+".availablePhysicalMemory")
+		}
+		if fields["memory_total"], ok = swapSpace["totalPhysicalMemory"].(float64); !ok {
+			return badFormatErr(te, swapSpace["totalPhysicalMemory"], "float64", swapSpaceMonitorRouter+".totalPhysicalMemory")
+		}
+	}
+	return nil
+}
+
+func (j *Jenkins) gatherNodesData(nodes []*gojenkins.Node, acc telegraf.Accumulator) {
+
+	tags := map[string]string{}
+	fields := make(map[string]interface{})
+	baseTe := typedErr{
+		url: j.URL + "/computer/api/json",
+	}
+
+	// get node data
+	for _, node := range nodes {
+		te := j.gatherNodeData(node, baseTe, fields, tags)
+		if te == nil {
+			acc.AddFields(measurementNode, fields, tags)
+			continue
+		}
+		switch te.level {
+		case continueLevel:
+			continue
+		default:
+			acc.AddError(te)
+		}
+	}
+}
+
+func (j *Jenkins) gatherJobs(jobNames []gojenkins.InnerJob, acc telegraf.Accumulator) {
+	jobsC := make(chan srcJob, j.MaxTCPConcurrentConnections)
+	errC := make(chan *typedErr)
+	var wg sync.WaitGroup
+	for _, job := range jobNames {
+		wg.Add(1)
+		go func(job gojenkins.InnerJob) {
+			jobsC <- srcJob{
+				name:    job.Name,
+				parents: []string{},
+				layer:   0,
+			}
+		}(job)
+	}
+
+	for i := 0; i < j.MaxTCPConcurrentConnections; i++ {
+		go j.getJobDetail(jobsC, errC, &wg, acc)
+	}
+
+	go func() {
+		wg.Wait()
+		close(errC)
+	}()
+
+	select {
+	case te := <-errC:
+		if te != nil {
+			acc.AddError(te)
+		}
+	}
+}
+
+func (j *Jenkins) getJobDetail(jobsC chan srcJob, errC chan<- *typedErr, wg *sync.WaitGroup, acc telegraf.Accumulator) {
+	for sj := range jobsC {
+		if j.MaxSubJobsLayer > 0 && sj.layer == j.MaxSubJobsLayer {
+			wg.Done()
```

Comment:
This is very hard to very the WaitGroup usage is correct, because the `Add` is not called at the front and not every return path calls `Done` and there is no use of defer.  You should call Add before starting the goroutine and using defer to ensure Done is called.